crates/sovereign-sdk/full-node/sov-stf-runner/src/runner.rs

```diff
@@ -735,353 +735,369 @@ where
                 .build();
             let inner_client = &self.sequencer_client;
-            let soft_batch = match retry_backoff(exponential_backoff.clone(), || async move {
-                match inner_client.get_soft_batch::<Da::Spec>(height).await {
-                    Ok(Some(soft_batch)) => Ok(soft_batch),
-                    Ok(None) => {
-                        debug!("Soft Batch: no batch at height {}, retrying...", height);
-
-                        // We wait for 2 seconds and then return a Permanent error so that we exit the retry.
-                        // This should not backoff exponentially
-                        sleep(Duration::from_secs(2)).await;
-                        Err(backoff::Error::Permanent(
-                            "No soft batch published".to_owned(),
-                        ))
-                    }
-                    Err(e) => match e.downcast_ref::<JsonrpseeError>() {
-                        Some(JsonrpseeError::Transport(e)) => {
-                            let error_msg =
-                                format!("Soft Batch: connection error during RPC call: {:?}", e);
-                            debug!(error_msg);
-                            Err(backoff::Error::Transient {
-                                err: error_msg,
-                                retry_after: None,
-                            })
-                        }
-                        _ => Err(backoff::Error::Transient {
-                            err: format!("Soft Batch: unknown error from RPC call: {:?}", e),
-                            retry_after: None,
-                        }),
-                    },
-                }
-            })
-            .await
-            {
-                Ok(soft_batch) => soft_batch,
-                Err(_) => {
-                    continue;
-                }
-            };
-
-            if last_l1_height != soft_batch.da_slot_height || cur_l1_block.is_none() {
-                last_l1_height = soft_batch.da_slot_height;
-                // TODO: for a node, the da block at slot_height might not have been finalized yet
-                // should wait for it to be finalized
-                let da_service = &self.da_service;
-                let filtered_block = retry_backoff(exponential_backoff.clone(), || async {
-                    da_service
-                        .get_block_at(soft_batch.da_slot_height)
+            let soft_batches: Vec<GetSoftBatchResponse> =
+                match retry_backoff(exponential_backoff.clone(), || async move {
+                    match inner_client
+                        .get_soft_batch_range::<Da::Spec>(height..height + 10)
                         .await
-                        .map_err(backoff::Error::transient)
+                    {
+                        Ok(soft_batches) => {
+                            Ok(soft_batches.into_iter().flatten().collect::<Vec<_>>())
```

Comment:
@eyusufatik This is where `null` values are filtered out for blocks we've requested that don't exist.